### ISSUE TITLE
[26.08] Extended Megatron FSDP example: verify combinations & document caveats

### ISSUE THREAD

**Issue by cuichenx** [NVIDIA-NeMo/Megatron-Bridge#4246]
## Summary

Create an **extended Megatron FSDP example/tutorial** that verifies FSDP across the combinations users actually run (pretraining, finetuning, LoRA; dense + MoE + VLM; TP/PP/EP/CP; packed sequences; checkpointing) and **documents how to enable each, the caveats, and the combinations that do not work**.

## Motivation

Megatron FSDP is the practical FSDP path in Bridge today, but in-repo validation is thin. As of the current skill/doc:

- The only validated functional coverage is a **2-GPU pretrain smoke** (`tests/functional_tests/training/test_megatron_fsdp.py`: `test_fsdp_pretrain_basic`, `..._with_checkpoint`, `..._save_resume`).
- The doc explicitly says broad compatibility is **"code-supported first, not benchmark-proven for every combination"** and the skill `card.yaml` lists as open follow-ups: *"Validate TP/PP/CP/EP compatibility matrix explicitly"*, finetune/LoRA/VLM coverage, and DDP-vs-distopt-vs-FSDP benchmarking.

Users have no end-to-end example showing FSDP with finetuning, LoRA, MoE/EP, VLM, or packed-THD, nor a clear "what works / what doesn't" matrix. This issue tracks closing that gap.

### Runs to verify
- [ ] Pretraining
- [ ] Finetuning (SFT)
- [ ] LoRA / PEFT

### Models
- [ ] **Small MoE:** Qwen3 MoE 30B (exercises EP + FSDP, fits a small node count)
- [ ] **Large dense/MoE:** Llama-3.1 405B **or** DeepSeek-V3 (full-scale, multi-node; see "Right-sizing" below)
- [ ] **VLM (stretch):** Nemotron Omni or Qwen3.5-VL — verify FSDP shards the LM while the vision encoder behaves (frozen/unfrozen)

### Feature combinations to verify (note enablement + caveats for each)
- [ ] **TP / PP / EP / CP** — the 5D-parallelism matrix × {pretrain, SFT, LoRA}
- [ ] **Packed sequence (THD)** — incl. the CP finetuning rules (`calculate_per_token_loss=True`, `ddp.average_in_collective=False`, `pad_seq_to_mult`)
- [ ] **Checkpointing** — `fsdp_dtensor` save/load/resume; **resharding across layouts** (DCP claims to remap when TP/PP changes between runs — verify); **HF export** from an `fsdp_dtensor` checkpoint

### Suggested additional combinations to cover
*(please verify if these are relevant and note caveats)*
- [ ] **`data_parallel_sharding_strategy` variants:** `optim_grads_params` vs `optim_grads` vs `optim` — which work, and the memory/throughput tradeoff
- [ ] **Mixed precision:** BF16 vs FP16, and **FP8** (note `keep_fp8_transpose_cache=False` is force-set in the perf harness — document why and the interaction); NVFP4/MXFP8 on Blackwell if feasible
- [ ] **CPU offloading** with FSDP — documented constraints are `PP==1` and activation recompute disabled; confirm and document
- [ ] **Activation recomputation / checkpointing** interplay with FSDP
- [ ] **Comm-overlap knobs:** `overlap_grad_reduce`, `overlap_param_gather` — applicability/benefit under FSDP's on-demand all-gather
- [ ] **Optimizers:** Adam vs Muon under FSDP (+ EP)
- [ ] **Multi-node caveat:** `CUDA_DEVICE_MAX_CONNECTIONS` — upstream warns FSDP and TP/CP can want different values on Hopper and earlier
- [ ] **Resiliency:** fault tolerance / in-process restart with `fsdp_dtensor` checkpoints
- [ ] **FSDP2 status:** confirm whether `use_torch_fsdp2` is still runtime-blocked (skill records a `pg_collection` `TypeError`); explicitly mark unsupported combos rather than leaving them silent

## Already-known constraints (don't re-discover — confirm & extend)

From `docs/training/megatron-fsdp.md` + `skills/perf-megatron-fsdp/`:
- Megatron FSDP requires `ckpt_format="fsdp_dtensor"`; **not interchangeable** with `torch_dist`/`zarr` (can't load FSDP ckpt into a non-FSDP run or vice-versa).
- `use_megatron_fsdp` and `use_torch_fsdp2` are **mutually exclusive**.
- `use_tp_pp_dp_mapping=True` is **not supported** with Megatron FSDP.
- Bridge auto-disables `average_in_collective` and some buffer-reuse opts under FSDP.
- Public recipes often expose `use_megatron_fsdp` but still default to `torch_dist` — recipe ergonomics lag the requirement.

**Minimal enablement (baseline for the tutorial):**
```python
cfg.dist.use_megatron_fsdp = True
cfg.ddp.use_megatron_fsdp = True
cfg.ddp.data_parallel_sharding_strategy = "optim_grads_params"
cfg.ddp.average_in_collective = False
cfg.checkpoint.ckpt_format = "fsdp_dtensor"
```

## Right-sizing for CI vs. full-scale

405B / DSv3 full runs are multi-node one-offs and can't live in CI. Suggest:
- **CI-affordable** (≤2 GPUs, per the functional-test budget): SFT, LoRA, packed-THD, a small TP/EP case on a toy/30B-proxy — added to `test_megatron_fsdp.py`.
- **Full-scale** (405B/DSv3/VLM): run once, capture loss curves + memory/throughput, record results in the doc/skill rather than as a recurring test.

## Deliverables
- [ ] **Tutorial script** (e.g. `tutorials/fsdp/…` or `examples/models/.../slurm_*.sh`) covering a representative subset of the above
- [ ] **Compatibility matrix** filled in (see table below) added to `docs/training/megatron-fsdp.md`
- [ ] **Update `docs/training/megatron-fsdp.md`** with verified enablement + caveats
- [ ] **Update `skills/perf-megatron-fsdp/`** (`SKILL.md` + `card.yaml` `validation_status`/`follow_up_validation`) with findings
- [ ] Extend `tests/functional_tests/training/test_megatron_fsdp.py` with the CI-affordable combos

### Compatibility matrix to fill in
| Combination | Pretrain | SFT | LoRA | Enablement / caveats |
|---|---|---|---|---|
| FSDP baseline (DP only) | | | | |
| FSDP + TP | | | | |
| FSDP + PP | | | | |
| FSDP + CP | | | | |
| FSDP + EP (MoE) | | | | |
| FSDP + packed THD | | | | |
| FSDP + checkpoint save/resume | | | | |
| FSDP + ckpt reshard (TP/PP change) | | | | |
| FSDP + FP8 | | | | |
| FSDP + CPU offload | | | | |
| FSDP + VLM | | | | |

Legend: ✅ works / ⚠️ works-with-caveat / ❌ unsupported (link evidence).

## Acceptance criteria
- Each checked combination is marked ✅/⚠️/❌ with evidence (test, run log, or code anchor).
- Every ⚠️/❌ has a documented reason and, where relevant, a workaround.
- Doc + skill updated; CI-affordable cases land as functional tests.

## References
- `docs/training/megatron-fsdp.md`
- `skills/perf-megatron-fsdp/SKILL.md`, `skills/perf-megatron-fsdp/card.yaml`
- `tests/functional_tests/training/test_megatron_fsdp.py`
- `src/megatron/bridge/training/config.py` (FSDP knobs + validation), `src/megatron/bridge/models/common/unimodal.py` (wrapper selection), `scripts/performance/utils/overrides.py` (perf harness FSDP overrides)

**Comment by gautham-kollu** [comment]
Adding FSDP/HSDP example for nemotron3 nano #4745
